apps/web/src/utils/trpc/QueryProvider.tsx

```diff
@@ -4,42 +4,81 @@ import { env } from "@/env"
+} from "@trpc/client"
+import { createContext, type Dispatch, type PropsWithChildren, type SetStateAction, useContext, useState } from "react"
 import superjson from "superjson"
 import { TRPCProvider } from "./client"
+import { secondsToMilliseconds } from "date-fns"
+
+// connecting is default, pending is when it is open, and idle idk
+export type TRPCSSEConnectionState = "connecting" | "pending" | "idle"
+
+export interface TRPCSSERegisterChangeConnectionStateContextType {
+  trpcSSERegisterChangeConnectionState: TRPCSSEConnectionState
+  setTRPCSSERegisterChangeConnectionState: Dispatch<SetStateAction<TRPCSSEConnectionState>>
+}
+
+const TRPCSSERegisterChangeConnectionStateContext =
+  createContext<TRPCSSERegisterChangeConnectionStateContextType | null>(null)
+
+export const useTRPCSSERegisterChangeConnectionState = () => {
+  const context = useContext(TRPCSSERegisterChangeConnectionStateContext)
+  if (!context) {
+    throw new Error("useTRPCSSERegisterChangeConnectionState() can only be used inside of a <TRPCProvider>")
+  }
+  return context
+}
 
 export const QueryProvider = ({ children }: PropsWithChildren) => {
   const session = useSession()
+  const [trpcSSERegisterChangeConnectionState, setTRPCSSERegisterChangeConnectionState] =
+    useState<TRPCSSEConnectionState>("connecting")
 
   const trpcConfig: CreateTRPCClientOptions<AppRouter> = {
     links: [
       loggerLink({
         enabled: (opts) =>
           env.NEXT_PUBLIC_ORIGIN === "development" || (opts.direction === "down" && opts.result instanceof Error),
       }),
-      httpBatchLink({
-        transformer: superjson,
-        url: `${env.NEXT_PUBLIC_RPC_HOST}/api/trpc`,
-        async fetch(url, options) {
-          const headers = new Headers(options?.headers)
-
-          if (session !== null) {
-            headers.append("Authorization", `Bearer ${session.accessToken}`)
-          }
-          return fetch(url, {
-            ...options,
-            credentials: "include",
-            headers,
-          })
-        },
+      splitLink({
+        condition: (op) => op.type === "subscription",
+        true: httpSubscriptionLink({
+          transformer: superjson,
+          url: `${env.NEXT_PUBLIC_RPC_HOST}/api/trpc`,
+        }),
+        false: httpBatchLink({
+          transformer: superjson,
+          url: `${env.NEXT_PUBLIC_RPC_HOST}/api/trpc`,
+          async fetch(url, options) {
+            const headers = new Headers(options?.headers)
+
+            if (session !== null) {
+              headers.append("Authorization", `Bearer ${session.accessToken}`)
+            }
+            return fetch(url, {
+              ...options,
+              credentials: "include",
+              headers,
+            })
+          },
+        }),
       }),
     ],
   }
   const [queryClient] = useState(
     () =>
       new QueryClient({
         defaultOptions: {
+          queries: {
+            staleTime: secondsToMilliseconds(30),
+          },
```

Comment:
Is this for all queries? Does this make every single query refetch every 30 seconds?

Reply:
No, by default the staleTime is zero. This means that each subscriber to a query will refetch it (after deduplication) as data is considered stale by default.

This change means it's going not going to refetch more often than every 30 seconds, unless the query is invalidated by us

https://tanstack.com/query/v4/docs/framework/react/guides/important-defaults